src/IROperator.h

```diff
@@ -327,6 +327,41 @@ Expr requirement_failed_error(Expr condition, const std::vector<Expr> &args);
 
 Expr memoize_tag_helper(Expr result, const std::vector<Expr> &cache_key_values);
 
+/** Cast a to the same type with twice the bits. */
+Expr widen(Expr a);
+/** Cast a to the same type with half the bits. */
+Expr narrow(Expr a);
+
+/** Compute widen(a) + widen(b). The result is always signed. */
+Expr widening_add(Expr a, Expr b);
+/** Compute widen(a) * widen(b). a and b may have different signedness. */
+Expr widening_mul(Expr a, Expr b);
+/** Compute widen(a) - widen(b). The result is always signed. */
+Expr widening_sub(Expr a, Expr b);
+/** Compute widen(a) << b. */
+Expr widening_shift_left(Expr a, Expr b);
+/** Compute widen(a) >> b. */
+Expr widening_shift_right(Expr a, Expr b);
+
+/** Compute (a + (1 >> min(b, 0)) / 2) << b */
```

Comment:
Is the comment right? That expression is quite confusing.

Reply:
I think it is! Note this is a rounding shift *left*, i.e. it's equivalent to `rounding_shift_right(a, -b)`.

If you believe rounding_shift_right is correct, then you can get from that one to this by relatively straightforward simplifications:

`rounding_right_shift(a, b)` = `(a + (1 << max(-b, 0)) / 2) >> -b` = `(a + (1 << -min(b, 0)) / 2) << b` = `(a + (1 >> min(b, 0)) / 2) << b` = `rounding_left_shift(a, b)`